### ISSUE TITLE
Applying Hamburger to other models makes the training collapse soon

### ISSUE THREAD

**Issue by npurson** [Gsunshine/Enjoy-Hamburger#10]
I've been working on applying Hamburger to other detection models (to be specific, Mask2Former & SparseInst), mainly by inserting Hamburger after the neck to align the multi-scale features, but the training process always collapses after only several iterations because of the nan output.
Given that the training recipe is rather general, and further reducing the lr does no help, I guess this indicates the gradient propagation is unstable? (p.s. applying `@torch.no_grad()` to `local_inference()` is also unhelpful)
Thus I'm wondering what's the intrinsic cause for this? have you ever met similar cases? or any suggestions for a fix? 

Any idea would be appreciated.

**Comment by Gsunshine** [comment]
Hi @npurson ,

Thank you for your interest in Hamburger!
This is a good question. If you check the details of Mask2Former or SparseInst, you might find that they essentially follow some similar ideas that try to build a semantic dictionary (using whatever queries or conv kernels) and coefficient (mask) matrix, but in an implicit manner, while Hamburger is explicitly doing this using optimization.

In all the training for semantic segmentation, we do not observe instability. But I do not know what happened in your case, especially if applying dictionary learning/decomposition twice, by combining the optimization-based methods (like Hamburger) and attention-based methods (Mask2Former). 

You can start your debugging by using a full gradient in NMF. Do not run full gradient for VQ, which might cause severe instability (but still not nan in my previous tests).

One thing worth noting is that you should check your input to NMF to make sure they are all non-negative values. I met the nan output value only at the very beginning stage of this work, maybe three years ago, when I first wrote NMF and tried to input representations into NMF but forgot to do the ReLU stuff.

Other reasons for nan might be from the attention part, which its training can be influenced by MD. Or the detection loss, in which we are using simple loss in semantic segmentation. The location in the network to use Hamburger/self-attention is also important. 

Overall, I would say Mask2Former or SparseInst already have some overlaps with Hamburger, at least partially and implicitly doing this. They are not orthogonal directions, so hard to determine how much we can improve by stacking them. You could try to devise new methods based on decomposition for detection instead of directly applying Hamburger.

Thanks!

Zhengyang

**Comment by npurson** [comment]
Indicating and helpful! Thanks for reaching out to me.
It turns out to be the problem that I also forgot to do the ReLU stuff : )
